2017 Robot/src/org/usfirst/frc/team4999/lights/Animator.java

```diff
@@ -6,106 +6,97 @@
 
 import edu.wpi.first.wpilibj.Timer;
 
+/**
+ * Runs in an infinite loop. Displays a frame of {@link Animation}, then waits the duration specified by the animation
+ * @author jordan
+ *
+ */
 class AnimatorThread extends Thread {
-	private Color[] currentState;
 	private Display out;
 	private Animation current;
-	private double brightness;
 	
-	private final int TIME_TO_SEND_FRAME = 50;
+	private final long WAIT_TIMEOUT = 500;
 	
+	// Generic object. The wait() method is used to pause the run thread until the notifyAll() method is called
+	// If an animation returns -1 for the time to wait, the run thread pauses execution until a new animation is set
+	private final Object pauseLock = new Object();
 	
-	public AnimatorThread(Display out, Color[] currentState, Animation current, double brightness) {
-		if(brightness < 0 || brightness > 1) throw new IllegalArgumentException("Brightness must be within [0,1]"); 
-		this.currentState = currentState;
+	public AnimatorThread(Display out, Animation current) {
 		this.out = out;
 		this.current = current;
-		this.brightness = brightness;
-		
-		this.setDaemon(true);
-	}
-	
-	private Color dim(Color in) {
-		return new Color((int)(in.getRed() * brightness), (int)(in.getGreen() * brightness), (int)(in.getBlue() * brightness));
 	}
 	
-	private Color[] setBrightness(Color[] in) {
-		Color[] out = in.clone();
-		for(int i = 0; i < out.length; i++) {
-			out[i] = dim(out[i]);
+	public void setAnimation(Animation newAnimation) {
+		this.current = newAnimation;
+		
+		// Tells the thread to continue. 
+		synchronized (pauseLock) {
+			pauseLock.notifyAll();
 		}
-		return out;
 	}
 	
 	public void run() {
-		while(!Thread.interrupted()) {
-			currentState = current.animate(currentState);
-			out.show(setBrightness(currentState));
-			int delay = current.getDelayUntilNextFrame();
-			//System.out.println("Expected: " + delay);
-			delay -= TIME_TO_SEND_FRAME;
-			delay = (delay < 0) ? 0 : delay;
+		while(!Thread.interrupted()){
+			// Note how long the send takes
+			long millis = System.currentTimeMillis();
+			Animation animation = current;
+			// show current frame
+			out.show(animation.getNextFrame());
+			// get how long to delay for
+			int delay = animation.getFrameDelayMilliseconds();
+			
+			synchronized (pauseLock) {
+				if(delay < 0) {
+					try {
+						pauseLock.wait(WAIT_TIMEOUT);
```

Comment:
Yeah, that makes sense. I hadn't considered the benefit of a refresh in case of a transmission error. I'll put a constant delay on the solid class and have it refresh every half-second or so